src/Features/Core/Portable/CodeRefactorings/AbstractRefactoringHelpersService.cs

```diff
@@ -2,34 +2,28 @@
 
 using System;
 using System.Collections.Generic;
+using System.Collections.Immutable;
 using System.Linq;
 using System.Threading;
 using System.Threading.Tasks;
 using Microsoft.CodeAnalysis.LanguageServices;
+using Microsoft.CodeAnalysis.PooledObjects;
 using Microsoft.CodeAnalysis.Shared.Extensions;
 using Microsoft.CodeAnalysis.Text;
+using Roslyn.Utilities;
 
 namespace Microsoft.CodeAnalysis.CodeRefactorings
 {
-    internal abstract class AbstractRefactoringHelpersService : IRefactoringHelpersService
+    internal abstract class AbstractRefactoringHelpersService<TExpressionSyntax, TArgumentSyntax> : IRefactoringHelpersService
+        where TExpressionSyntax : SyntaxNode
+        where TArgumentSyntax : SyntaxNode
     {
-        public async Task<TSyntaxNode> TryGetSelectedNodeAsync<TSyntaxNode>(
-            Document document, TextSpan selection, CancellationToken cancellationToken) where TSyntaxNode : SyntaxNode
-        {
-            return (TSyntaxNode)await TryGetSelectedNodeAsync(
-                document,
-                selection, n => n is TSyntaxNode,
-                ExtractNodesSimple, ExtractNodesIfInHeader,
-                cancellationToken).ConfigureAwait(false);
-        }
-
-        protected async Task<SyntaxNode> TryGetSelectedNodeAsync(
+        public async Task<ImmutableArray<TSyntaxNode>> GetRelevantNodesAsync<TSyntaxNode>(
             Document document, TextSpan selectionRaw,
-            Func<SyntaxNode, bool> predicate,
-            Func<SyntaxNode, ISyntaxFactsService, IEnumerable<SyntaxNode>> extractNodes,
-            Func<SyntaxNode, int, ISyntaxFactsService, IEnumerable<SyntaxNode>> extracNodestIfInHeader,
-            CancellationToken cancellationToken)
+            CancellationToken cancellationToken) where TSyntaxNode : SyntaxNode
         {
+            var relevantNodesBuilder = ArrayBuilder<TSyntaxNode>.GetInstance();
```

Comment:
leaked on the first return below.  consider moving tdown below that.

Reply:
one simple way to structure code to avoid this is to do the following:

```c#
public async Task<ImmutableArray<TSyntaxNode>> GetRelevantNodesAsync<TSyntaxNode>(...)
{
    var builder = ArrayBuilder.GetInstance();
    try
    {
        AddRelevantNodesAsync(..., builder);
        return builder.ToImmutable();
    }
    finally
    {
        builder.free();
    }
}
```

Reply:
this also gives you a good place where you can order the results first by size.

Reply:
Speaking of which, what's the reason it doesn't implement Disposable & therefore Using statement?

Reply:
yeah, i'd like it to do that :D

Reply:
I'll try to look into that in separate PR.